cpp/src/cluster/detail/mst.cuh

```diff
@@ -117,6 +118,149 @@ void connect_knn_graph(
   merge_msts<value_idx, value_t>(msf, new_mst, stream);
 }
 
+/**
+ * Connect an unconnected knn graph (one in which mst returns an msf). The
+ * device buffers underlying the Graph_COO object are modified in-place.
+ * @tparam value_idx index type
+ * @tparam value_t floating-point value type
+ * @param[in] handle raft handle
+ * @param[in] X original dense data on host memory from which knn graph was constructed
+ * @param[inout] msf edge list containing the mst result
+ * @param[in] m number of rows in X
+ * @param[in] n number of columns in X
+ * @param[in] n_components number of components in color
+ * @param[inout] color the color labels array returned from the mst invocation
+ * @return updated MST edge list
+ */
+template <typename value_idx, typename value_t>
+void connect_knn_graph(
+  raft::resources const& handle,
+  const value_t* X,
+  raft::sparse::solver::Graph_COO<value_idx, value_idx, value_t>& msf,
+  size_t m,
+  size_t n,
+  int n_components,
+  value_idx* color,
+  cuvs::distance::DistanceType metric = cuvs::distance::DistanceType::L2SqrtExpanded)
+{
+  auto stream = raft::resource::get_cuda_stream(handle);
+
+  // Copy color array from device to host
+  std::vector<value_idx> h_color(m);
+  raft::copy(h_color.data(), color, m, stream);
+  raft::resource::sync_stream(handle, stream);
+
+  std::unordered_map<value_idx, value_idx> color_remap;
+  value_idx new_label = 0;
+
+  // Build remapping table so that colors are compact integers (i.e. consecutive colors)
+  for (size_t i = 0; i < m; ++i) {
+    if (color_remap.find(h_color[i]) == color_remap.end()) {
+      color_remap[h_color[i]] = new_label++;
+    }
+  }
+
+  for (size_t i = 0; i < m; ++i) {
+    h_color[i] = color_remap[h_color[i]];
+  }
+
+  // make key (color) : value (vector of ids that have that color)
+  std::unordered_map<value_idx, std::vector<value_idx>> component_map;
+  for (value_idx i = 0; i < static_cast<value_idx>(m); ++i) {
+    component_map[h_color[i]].push_back(i);
+  }
+
+  std::vector<std::tuple<value_idx, value_idx, value_t>> selected_edges;
+
+  std::random_device rd;
+  std::mt19937 gen(rd());
+  std::uniform_int_distribution<> dis;
+
+  auto pairwise_dist = raft::make_device_matrix<value_t, int64_t>(handle, 1, 1);
+  auto data_u        = raft::make_device_matrix<value_t, int64_t>(handle, 1, n);
+  auto data_v        = raft::make_device_matrix<value_t, int64_t>(handle, 1, n);
+
+  // connect i-1 component and i component
+  for (int i = 1; i < n_components; ++i) {
```

Comment:
How many components are there usually? If the number is on the larger side, this could benefit with some omp parallelism

Reply:
very small. Based on heuristic observations, it's usually somewhere between 1 ~50.

Reply:
Okay, so will this be a if-else omp launch?
Also, I just wanted to confirm that we would want to use a separate stream per omp thread so that we can launch kernels concurrently?

Reply:
Not an if-else, we can assign the number of threads based on some condition before the for loop (n_components / 5)?

And yes, 1 stream per thread.

Reply:
@divyegala ,thinking of it, if make this an omp region, we have to have `data_u`, `data_v`, and `pairwise_dist` matrix allocation inside the for loop. do you think this will be too much of an overhead?

maybe just gathering all vectors into a matrix and calling the pariwise dist kernel once will be a better idea?

Reply:
Yes, gather first is 100% better. For some reason I thought doing this with a for loop was a requirement.

Memory overhead is only (n_components, n_cols) and (n_components, n_components) where n_components ~ 50.